File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        curr1 = list1
        curr2= list2
        if curr1 == None:
            return curr2
        if curr2 == None:
            return curr1
        if  curr1.val <=curr2.val :
            main = curr1
            curr1 = curr1.next
        else:
            main = curr2
            curr2= curr2.next
        og = main
        while curr1 !=None and curr2 != None:
            if curr1.val <=curr2.val:
                main.next = curr1
                main = main.next
                curr1 = curr1.next
            else:
                main.next = curr2
                main = main.next
                curr2 = curr2.next

        if curr1:
            main.next = curr1
        else:
            main.next = curr2
        return og

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode* og = new ListNode();
        ListNode* tail = og;

        
        while (true){
            int minIdx = -1;
            for (int i = 0; i < lists.size(); i +=1){
                if (lists[i] != nullptr && (minIdx == -1 ||lists[i]->val < lists[minIdx]->val)){
                    minIdx = i;
                }

            }
            if (minIdx == -1){break;}
            tail->next = lists[minIdx];
            tail = tail->next;
            lists[minIdx] = lists[minIdx]->next; 
        }
        return og->next;
    }
};
